bin/data_frame_analyzer/Main.cc

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+//! \brief
+//! Applies a range of ML analyses on a data frame.
+//!
+//! DESCRIPTION:\n
+//! Expects to be streamed CSV or length encoded data on STDIN or a named pipe,
+//! and sends its JSON results to STDOUT or another named pipe.
+//!
+//! IMPLEMENTATION DECISIONS:\n
+//! Standalone program.
+//!
+#include <core/CJsonOutputStreamWrapper.h>
+#include <core/CLogger.h>
+#include <core/CProcessPriority.h>
+#include <core/CoreTypes.h>
+
+#include <ver/CBuildInfo.h>
+
+#include <api/CDataFrameAnalyzer.h>
+#include <api/CIoManager.h>
+#include <api/CLengthEncodedInputParser.h>
+#include <api/COutputChainer.h>
+
+#include "CCmdLineParser.h"
+
+#include <boost/bind.hpp>
+#include <boost/scoped_ptr.hpp>
+
+#include <string>
+
+#include <stdlib.h>
+
+int main(int argc, char** argv) {
+    // Read command line options
+    std::string configFile;
+    std::string logProperties;
+    std::string logPipe;
+    std::string inputFileName;
+    bool isInputFileNamedPipe(false);
+    std::string outputFileName;
+    bool isOutputFileNamedPipe(false);
+    if (ml::data_frame_analyzer::CCmdLineParser::parse(
+            argc, argv, configFile, logProperties, logPipe, inputFileName,
+            isInputFileNamedPipe, outputFileName, isOutputFileNamedPipe) == false) {
+        return EXIT_FAILURE;
+    }
+
+    // Construct the IO manager before reconfiguring the logger, as it performs
+    // std::ios actions that only work before first use
+    ml::api::CIoManager ioMgr(inputFileName, isInputFileNamedPipe,
+                              outputFileName, isOutputFileNamedPipe);
+
+    if (ml::core::CLogger::instance().reconfigure(logPipe, logProperties) == false) {
+        LOG_FATAL("Could not reconfigure logging");
+        return EXIT_FAILURE;
+    }
+
+    // Log the program version immediately after reconfiguring the logger.  This
+    // must be done from the program, and NOT a shared library, as each program
+    // statically links its own version library.
+    LOG_DEBUG(<< ml::ver::CBuildInfo::fullInfo());
+
+    ml::core::CProcessPriority::reducePriority();
+
+    if (ioMgr.initIo() == false) {
+        LOG_FATAL("Failed to initialise IO");
+        return EXIT_FAILURE;
+    }
+
+    typedef boost::scoped_ptr<ml::api::CInputParser> TScopedInputParserP;
+    TScopedInputParserP inputParser;
+    inputParser.reset(new ml::api::CLengthEncodedInputParser(ioMgr.inputStream()));
```

Comment:
Two things I am confused about:

* Why the `typedef` ? Are you expecting to use this type again later or is this strictly for line length purposes?
* Why not use `std::unique_ptr`? I don't understand why `boost::scoped_ptr` is being used in this instance.

Reply:
All very good questions! This is copy-pasta in all our `Main.cc` :P. I'll look into your points and refactor if possible

Reply:
The code this was copied from is using `boost::scoped_ptr` because it predates move semantics. You can use `boost::make_unique` to make a `std::unique_ptr` and store the result in an `auto` variable.

Reply:
So why wouldn't the following work? I suppose it could be considered harder to read, but it does not use any move semantics to my knowledge. 
```
const std::unique_ptr<ml::api::CInputParser> inputParser(new ml::api::CLengthEncodedInputParser(ioMgr.inputStream()));
```

Reply:
It would work, but there are [reasons](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3588.txt) to prefer make_unique.

Reply:
(It is planned to gradually phase out direct calls to new from the code base now we can have factory methods return unique_ptr.)

Reply:
OK, cool, wanted to make sure I was not crazy :).